### ISSUE TITLE
CBLK: Write/read request time out when being used in parallel

### ISSUE THREAD

**Issue by fhaverkamp** [open-power/snap#492]
I am unsure if this is a hardware or a software problem. In case of Eberhards test-case with caching setup to a minimum, I see write but also reads not coming to an end. Let's have a look together to see what this might be caused by.

```
B 0000baa3.0000bb18 1509375734639    [block_write] writing (0x3ffda90ae810 LBA=195142964 nblocks=2) ...
B 0000baa3.0000baa4 1509375734655    [block_read] reading (0x3ffda90ac810 LBA=195011166 nblocks=2) ...
B 0000baa3.0000bb18 1509375734655    [get_write_req] GIVE OUT WRITE slot 0 LBA=195142964
B 0000baa3.0000bb18 1509375734682        [start_memcpy] HW COPY_HN memcpy_3(slot=0 dest=0x5d0d29a0, src=0x3ffda9dc0000 n=8192 bytes) LBA=195142964
B 0000baa3.0000baa4 1509375734684    [get_read_req] GIVE OUT READ slot 2 LBA=195011166
B 0000baa3.0000baa4 1509375734704        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd12f0 n=8192 bytes) LBA=195011166
B 0000baa3.0000bb14 1509375734804        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195011166
B 0000baa3.0000bb14 1509375734810      [completion_thread] waking up slot 2 LBA=195011166
B 0000baa3.0000baa4 1509375734831    [block_read] reading (0x3ffda90aa810 LBA=195011168 nblocks=2) ...
B 0000baa3.0000baa4 1509375734838    [get_read_req] GIVE OUT READ slot 2 LBA=195011168
B 0000baa3.0000baa4 1509375734844        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd1300 n=8192 bytes) LBA=195011168
B 0000baa3.0000bb14 1509375734947        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195011168
B 0000baa3.0000bb14 1509375734954      [completion_thread] waking up slot 2 LBA=195011168
B 0000baa3.0000baa4 1509375734973    [block_read] reading (0x3ffda90a6810 LBA=195011170 nblocks=2) ...
B 0000baa3.0000baa4 1509375734980    [get_read_req] GIVE OUT READ slot 2 LBA=195011170
B 0000baa3.0000baa4 1509375734986        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd1310 n=8192 bytes) LBA=195011170
B 0000baa3.0000bb14 1509375735086        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195011170
B 0000baa3.0000bb14 1509375735092      [completion_thread] waking up slot 2 LBA=195011170
B 0000baa3.0000baa4 1509375735109    [block_read] reading (0x3ffda909a810 LBA=195012384 nblocks=2) ...
B 0000baa3.0000baa4 1509375735115    [get_read_req] GIVE OUT READ slot 2 LBA=195012384
B 0000baa3.0000baa4 1509375735121        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd3900 n=8192 bytes) LBA=195012384
B 0000baa3.0000bb14 1509375735222        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195012384
B 0000baa3.0000bb14 1509375735228      [completion_thread] waking up slot 2 LBA=195012384
B 0000baa3.0000baa4 1509375735245    [block_read] reading (0x3ffda908e810 LBA=195013014 nblocks=2) ...
B 0000baa3.0000baa4 1509375735251    [get_read_req] GIVE OUT READ slot 2 LBA=195013014
B 0000baa3.0000baa4 1509375735257        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd4cb0 n=8192 bytes) LBA=195013014
B 0000baa3.0000bb14 1509375735355        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195013014
B 0000baa3.0000bb14 1509375735363      [completion_thread] waking up slot 2 LBA=195013014
B 0000baa3.0000baa4 1509375735383    [block_read] reading (0x3ffda9082810 LBA=195015298 nblocks=2) ...
B 0000baa3.0000baa4 1509375735389    [get_read_req] GIVE OUT READ slot 2 LBA=195015298
B 0000baa3.0000baa4 1509375735396        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd9410 n=8192 bytes) LBA=195015298
B 0000baa3.0000bb14 1509375735494        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195015298
B 0000baa3.0000bb14 1509375735499      [completion_thread] waking up slot 2 LBA=195015298
B 0000baa3.0000baa4 1509375735517    [block_read] reading (0x3ffda90ea810 LBA=195015300 nblocks=2) ...
B 0000baa3.0000baa4 1509375735522    [get_read_req] GIVE OUT READ slot 2 LBA=195015300
B 0000baa3.0000baa4 1509375735528        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd9420 n=8192 bytes) LBA=195015300
B 0000baa3.0000bb14 1509375735628        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195015300
B 0000baa3.0000bb14 1509375735633      [completion_thread] waking up slot 2 LBA=195015300
B 0000baa3.0000baa4 1509375735650    [block_read] reading (0x3ffda90d0810 LBA=195015320 nblocks=2) ...
B 0000baa3.0000baa4 1509375735656    [get_read_req] GIVE OUT READ slot 2 LBA=195015320
B 0000baa3.0000baa4 1509375735662        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfd94c0 n=8192 bytes) LBA=195015320
B 0000baa3.0000bb14 1509375735762        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195015320
B 0000baa3.0000bb14 1509375735767      [completion_thread] waking up slot 2 LBA=195015320
B 0000baa3.0000baa4 1509375735784    [block_read] reading (0x3ffda90c6810 LBA=195016644 nblocks=2) ...
B 0000baa3.0000baa4 1509375735789    [get_read_req] GIVE OUT READ slot 2 LBA=195016644
B 0000baa3.0000baa4 1509375735797        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfdbe20 n=8192 bytes) LBA=195016644
B 0000baa3.0000bb14 1509375735897        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195016644
B 0000baa3.0000bb14 1509375735901      [completion_thread] waking up slot 2 LBA=195016644
B 0000baa3.0000baa4 1509375735919    [block_read] reading (0x3ffda90c2810 LBA=195017778 nblocks=2) ...
B 0000baa3.0000baa4 1509375735925    [get_read_req] GIVE OUT READ slot 2 LBA=195017778
B 0000baa3.0000baa4 1509375735931        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfde190 n=8192 bytes) LBA=195017778
B 0000baa3.0000bb14 1509375736032        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195017778
B 0000baa3.0000bb14 1509375736037      [completion_thread] waking up slot 2 LBA=195017778
B 0000baa3.0000baa4 1509375736053    [block_read] reading (0x3ffda90b8810 LBA=195019440 nblocks=2) ...
B 0000baa3.0000baa4 1509375736060    [get_read_req] GIVE OUT READ slot 2 LBA=195019440
B 0000baa3.0000baa4 1509375736068        [start_memcpy] HW COPY_NH memcpy_204(slot=2 dest=0x3ffda9e00000, src=0x5cfe1580 n=8192 bytes) LBA=195019440
B 0000baa3.0000bb14 1509375736170        [completion_status] HW READ_COMPLETED 00000012 slot: 2 LBA=195019440
B 0000baa3.0000bb14 1509375736176      [completion_thread] waking up slot 2 LBA=195019440
B 0000baa3.0000baa4 1509375736198    [block_write] writing (0x3ffda909e810 LBA=195009570 nblocks=2) ...
B 0000baa3.0000bb14 1509375740675      err: req[ 0]: WRITING 5/6 TIMEOUT!
```

Testcase is Eberhards NVMe block test-case.

**Comment by fhaverkamp** [comment]
In contrast to my own test case this is a situation where we have reads and writes going on in parallel. Interestingly I saw areas in the test case where just one read and one write was doing in parallel, even though the test case config said that it could use 8 threads. So there seems to be some sequential stuff going on. Regarding the LBAs used, I do not see increasing or decreasing sequences on the first glance. Looks pretty random without automatic analysis applied to it.

**Comment by fhaverkamp** [comment]
Standalone test case with reads and writes in parallel. Shows same behaviour. Retries might help.

```
haver@tul3eth3:~/snap$ SNAP_TRACE=0x80 LD_LIBRARY_PATH=/home/haver/snap/actions/hdl_nvme_example/sw:/home/haver/snap/software/lib ./actions/hdl_nvme_example/sw/snap_cblk -C0 -n0x40000 -t16 -b1 -R42 --rw # 2> output.log 
NVMe device has 209715200 blocks of each 4096 bytes; 819200 MiB (MAX 0xc800000 blocks) @ 16 threads
Reading/writing 262144 times 4096 bytes: 1048576 KiB NVMe into snap_cblk.bin
[check_req_timeouts] err: req[ 0]: WRITING 5/6 sec LBA=7793 TIMEOUT!
[check_req_timeouts] err: req[ 0]: WRITING 5/6 sec LBA=48055 TIMEOUT!
[check_req_timeouts] err: req[ 9]: READING 5/6 sec LBA=95186 TIMEOUT!
[check_req_timeouts] err: req[ 0]: WRITING 5/6 sec LBA=130801 TIMEOUT!
[check_req_timeouts] err: req[ 7]: READING 5/6 sec LBA=225270 TIMEOUT!
[check_req_timeouts] err: req[ 2]: READING 5/6 sec LBA=230974 TIMEOUT!
[check_req_timeouts] err: req[ 0]: WRITING 5/6 sec LBA=243073 TIMEOUT!
MIN: 4294967295 usecs, MAX: 0 usecs, AVG: 0 usecs
Reading/writing of 1073741824 bytes with 16 threads took 49809391 usec @ 21.557 MiB/sec random ordering
S 00014d5e.00014d5e 1509630589688    Statistics
  prefetches:          0
  prefetch_collis_4k:  0
  cache_hits:          0
    cache_hits_4k:     0
  hw_block_reads:      131075
  hw_block_writes:     131076
  block_reads:         131072
    block_reads_4k:    131072
  block_writes:        131072
    block_writes_4k:   131072
  idle_wakeups:        16
  cache_trashing_4k:   130560
  running:             49956561 usec
  reading:             66626202 usec
  writing:             24980129 usec
  rbytes_total:        536883200 10.747 MiB/sec
  wbytes_total:        536887296 10.747 MiB/sec
  max_read_usecs:      26403 usec
  max_write_usecs:     23315 usec
  avg_read_usecs:      508 usec
  avg_write_usecs:     190 usec
  min_read_usecs:      80 usec
  min_write_usecs:     20 usec
  avg_hw_read_usecs:   501 usec
  avg_hw_write_usecs:  186 usec
S 00014d5e.00014d5e 1509630589745      req[ 0]: IDLE LBA=261879 45 usec err_total=4
S 00014d5e.00014d5e 1509630589751      req[ 1]: IDLE LBA=262114 84 usec err_total=0
S 00014d5e.00014d5e 1509630589757      req[ 2]: IDLE LBA=262116 88 usec err_total=1
S 00014d5e.00014d5e 1509630589763      req[ 3]: IDLE LBA=262118 85 usec err_total=0
S 00014d5e.00014d5e 1509630589769      req[ 4]: IDLE LBA=262120 88 usec err_total=0
S 00014d5e.00014d5e 1509630589775      req[ 5]: IDLE LBA=262122 88 usec err_total=0
S 00014d5e.00014d5e 1509630589780      req[ 6]: IDLE LBA=262124 88 usec err_total=0
S 00014d5e.00014d5e 1509630589786      req[ 7]: IDLE LBA=262126 82 usec err_total=1
S 00014d5e.00014d5e 1509630589792      req[ 8]: IDLE LBA=262128 90 usec err_total=0
S 00014d5e.00014d5e 1509630589797      req[ 9]: IDLE LBA=262130 86 usec err_total=1
S 00014d5e.00014d5e 1509630589803      req[10]: IDLE LBA=262132 89 usec err_total=0
S 00014d5e.00014d5e 1509630589809      req[11]: IDLE LBA=262134 86 usec err_total=0
S 00014d5e.00014d5e 1509630589814      req[12]: IDLE LBA=262136 90 usec err_total=0
S 00014d5e.00014d5e 1509630589820      req[13]: IDLE LBA=262138 88 usec err_total=0
S 00014d5e.00014d5e 1509630589826      req[14]: IDLE LBA=262140 90 usec err_total=0
S 00014d5e.00014d5e 1509630589831      req[15]: IDLE LBA=262142 92 usec err_total=0
```

**Comment by fhaverkamp** [comment]
Running on our testbed tul3.

**Comment by fhaverkamp** [comment]
```
haver@tul3eth3:~/snap$ ./software/tools/snap_maint -C0 -vvv
[main] Enter
[snap_open] Enter: /dev/cxl/afu0.0m
[snap_open] Exit 0x1000c4d0420
[snap_version] Enter
SNAP on N250S Card, NVME enabled, 4096 MB SRAM available.
SNAP FPGA Release: v1.1.1 Distance: 44 GIT: 0xb8d81fd7
SNAP FPGA Build (Y/M/D): 2017/10/05 Time (H:M): 17:41
SNAP FPGA CIR Master: 1 My ID: 0
SNAP FPGA Up Time: 508493 sec
[snap_version] Exit
[snap_m_init] Enter
SNAP FPGA Exploration already done (MSAT: 1 MAID: 1)

   Short |  Action Type |   Level   |
   ------+--------------+-----------+-----------
     0     0x10140001     0x00000000  IBM HDL NVMe example

[snap_m_init] Exit rc: 0
[main] Exit rc: 0
[snap_close] Enter
[snap_close] Exit 0
haver@tul3eth3:~/snap$ 
```

**Comment by fhaverkamp** [comment]
```
haver@tul3eth3:~/snap$ ./actions/hdl_nvme_example/sw/snap_cblk -C0 --read -t10 -b2 -n0x80000 snap_read.bin
NVMe device has 209715200 blocks of each 4096 bytes; 819200 MiB (MAX 0xc800000 blocks) @ 10 threads
Reading 262144 times 8192 bytes: 2097152 KiB NVMe into snap_read.bin
[check_req_timeouts] err: req[ 1]: READING 3/4 sec LBA=0 TIMEOUT
[check_req_timeouts] err: req[ 8]: READING 3/4 sec LBA=14 TIMEOUT
[check_req_timeouts] err: req[10]: READING 3/4 sec LBA=18 TIMEOUT
[check_req_timeouts] err: req[12]: READING 3/4 sec LBA=22 TIMEOUT
[check_req_timeouts] err: req[13]: READING 3/4 sec LBA=24 TIMEOUT
[check_req_timeouts] err: req[14]: READING 3/4 sec LBA=26 TIMEOUT
[check_req_timeouts] err: req[ 6]: READING 3/4 sec LBA=34 TIMEOUT
[check_req_timeouts] err: req[ 7]: READING 3/4 sec LBA=36 TIMEOUT
[check_req_timeouts] err: req[ 9]: READING 3/4 sec LBA=38 TIMEOUT
[check_req_timeouts] err: req[11]: READING 3/4 sec LBA=40 TIMEOUT
[check_req_timeouts] err: req[ 1]: READING 3/4 sec LBA=0 TIMEOUT
[check_req_timeouts] err: req[ 8]: READING 3/4 sec LBA=14 TIMEOUT
[check_req_timeouts] err: req[10]: READING 3/4 sec LBA=18 TIMEOUT
[check_req_timeouts] err: req[12]: READING 3/4 sec LBA=22 TIMEOUT
[check_req_timeouts] err: req[13]: READING 3/4 sec LBA=24 TIMEOUT
[check_req_timeouts] err: req[14]: READING 3/4 sec LBA=26 TIMEOUT
[check_req_timeouts] err: req[ 6]: READING 3/4 sec LBA=34 TIMEOUT
[check_req_timeouts] err: req[ 7]: READING 3/4 sec LBA=36 TIMEOUT
[check_req_timeouts] err: req[ 9]: READING 3/4 sec LBA=38 TIMEOUT
[check_req_timeouts] err: req[11]: READING 3/4 sec LBA=40 TIMEOUT
[check_req_timeouts] err: req[ 1]: READING 3/4 sec LBA=0 TIMEOUT
D [cache_unreserve] warn: 0x3fff86410530 forcing status LBA=0/0 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff864107e0 forcing status LBA=1/1 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
[check_req_timeouts] err: req[ 8]: READING 3/4 sec LBA=14 TIMEOUT
[check_req_timeouts] err: req[10]: READING 3/4 sec LBA=18 TIMEOUT
D [cache_unreserve] warn: 0x3fff86412ad0 forcing status LBA=14/14 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86412d80 forcing status LBA=15/15 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
D [cache_unreserve] warn: 0x3fff86413590 forcing status LBA=18/18 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86413840 forcing status LBA=19/19 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
[check_req_timeouts] err: req[12]: READING 3/4 sec LBA=22 TIMEOUT
D [cache_unreserve] warn: 0x3fff86414050 forcing status LBA=22/22 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86414300 forcing status LBA=23/23 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
[check_req_timeouts] err: req[13]: READING 3/4 sec LBA=24 TIMEOUT
D [cache_unreserve] warn: 0x3fff864145b0 forcing status LBA=24/24 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86414860 forcing status LBA=25/25 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
[check_req_timeouts] err: req[14]: READING 3/4 sec LBA=26 TIMEOUT
D [cache_unreserve] warn: 0x3fff86414b10 forcing status LBA=26/26 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86414dc0 forcing status LBA=27/27 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
[check_req_timeouts] err: req[ 6]: READING 3/4 sec LBA=34 TIMEOUT
D [cache_unreserve] warn: 0x3fff86416090 forcing status LBA=34/34 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86416340 forcing status LBA=35/35 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
[check_req_timeouts] err: req[ 7]: READING 3/4 sec LBA=36 TIMEOUT
[check_req_timeouts] err: req[ 9]: READING 3/4 sec LBA=38 TIMEOUT
D [cache_unreserve] warn: 0x3fff864165f0 forcing status LBA=36/36 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff864168a0 forcing status LBA=37/37 cache from READING to UNUSED!
[check_req_timeouts] err: req[11]: READING 3/4 sec LBA=40 TIMEOUT
err: cblk_read unhappy rc=0! Timer expired
D [cache_unreserve] warn: 0x3fff86416b50 forcing status LBA=38/38 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86416e00 forcing status LBA=39/39 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
D [cache_unreserve] warn: 0x3fff864170b0 forcing status LBA=40/40 cache from READING to UNUSED!
D [cache_unreserve] warn: 0x3fff86417360 forcing status LBA=41/41 cache from READING to UNUSED!
err: cblk_read unhappy rc=0! Timer expired
MIN: 66 usecs, MAX: 637 usecs, AVG: 0 usecs
Reading of 2147483648 bytes with 10 threads took 12001129 usec @ 178.940 MiB/sec linear ordering
haver@tul3eth3:~/snap$ ./actions/hdl_nvme_example/sw/snap_cblk -C0 --rw -t10 -b2 -n0x80000 snap_read.bin
NVMe device has 209715200 blocks of each 4096 bytes; 819200 MiB (MAX 0xc800000 blocks) @ 10 threads
Reading/writing 262144 times 8192 bytes: 2097152 KiB NVMe into snap_read.bin
[check_req_timeouts] err: req[ 0]: WRITING 3/4 sec LBA=0 TIMEOUT
[check_req_timeouts] err: req[ 0]: WRITING 3/4 sec LBA=0 TIMEOUT
[check_req_timeouts] err: req[ 0]: WRITING 3/4 sec LBA=0 TIMEOUT
haver@tul3eth3:~/snap$ 
```

**Comment by fhaverkamp** [comment]
I do not see failures with the latest code. One fix I needed in my read/format test case where I accidentally deleted the code which wrote the output file.

See here: https://github.com/open-power/snap/pull/569
